src/stream.rs

```diff
@@ -0,0 +1,533 @@
+{
+    let page_size = page_size::get();
+    if BLOCK_SIZE < page_size {
+        return Err(StreamError::PageSize(BLOCK_SIZE, page_size))
+    }
+    let buffer = Arc::new(Buffer::new(0)?);
+    let shared = Arc::new(Shared {
+        length: AtomicU64::from(0),
+        file: ArcSwapOption::empty(),
+        current_buffer: ArcSwap::new(buffer.clone()),
+    });
+    let writer = StreamWriter {
+        shared: shared.clone(),
+        length: 0,
+        buf: buffer.ptr,
+        ptr: buffer.ptr,
+        file: None,
+        spare_buffer: None,
+    };
+    let reader = StreamReader {
+        shared,
+        mappings: LruBTreeMap::new(MAP_CACHE_PER_READER),
+    };
+    Ok((writer, reader))
+}
+
+impl<const BLOCK_SIZE: usize> StreamWriter<BLOCK_SIZE> {
+    /// Get the current length of the stream, in bytes.
+    pub fn len(&self) -> u64 {
+        self.length
+    }
+
+    /// Append data to the end of the stream.
+    ///
+    /// Returns the new stream length.
+    ///
+    pub fn append(&mut self, mut data: &[u8]) -> Result<u64, StreamError> {
+        let length = data.len();
+        let buffered = self.length as usize & Self::block_mask();
+        if buffered + length <= Self::block_size() {
+            // All the data will fit in the existing buffer.
+            unsafe { self.write_to_buffer(data, length) };
+            if self.length as usize & Self::block_mask() == 0 {
+                // Buffer is now full, and can be written to file.
+                unsafe { self.write_buffer_to_file()? };
+            }
+        } else {
+            // The data will fill the existing buffer.
+            if buffered > 0 {
+                // The buffer is partly used. Fill it and write it out.
+                let length = Self::block_size() - buffered;
+                unsafe { self.write_to_buffer(data, length) };
+                // Buffer is now full, and can be written to file.
+                unsafe { self.write_buffer_to_file()? };
+                data = &data[length..];
+            }
+            // The buffer is curently empty, so we are free to write whole
+            // blocks of data directly to the file, bypassing the buffer.
+            let direct = data.len() & !Self::block_mask();
+            if direct > 0 {
+                unsafe { self.write_to_file(&data[..direct])? };
+                data = &data[direct..];
+                self.length += direct as u64;
+            }
+            let length = data.len();
+            if length > 0 {
+                // There is data remaining which will fit in the next buffer.
+                unsafe { self.write_to_buffer(data, length) };
+            }
+        }
+        // Update shared length value, and return new length.
+        self.shared.length.store(self.length, Release);
+        Ok(self.length)
+    }
+
+    /// Helper method for writing data to buffer.
+    ///
+    /// The data must be guaranteed to fit within the remaining space.
+    ///
```

Comment:
I recently noticed a section in the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/documentation.html) book that recommends adding an explicit "Safety" heading for documenting `unsafe` usage.

e.g. this would just become:

```
    /// Helper method for writing data to buffer.
    ///
    /// # Safety
    ///
    /// The data must be guaranteed to fit within the remaining space.
    ///
```

Reply:
Yes, it would probably be better to be more explicit here, both about the fact this is the safety condition, and also what code is responsible for ensuring it. I think we can do without the heading when the comment fits on one line, though:
```
/// Safety: The caller must ensure that the data will fit within the remaining space in the buffer.
```

Reply:
Cosmetically I'd argue for using a heading so the rendered docs stay in line with any of the other "official" doc sections - i.e.

![screenshot](https://user-images.githubusercontent.com/44808/231495653-49ab9077-81fe-4ddc-a213-394d9d2cd198.png)

Reply:
Oh I take that back, clippy has some:

https://rust-lang.github.io/rust-clippy/rust-1.68.0/index.html#missing_errors_doc
https://rust-lang.github.io/rust-clippy/rust-1.68.0/index.html#missing_panics_doc
https://rust-lang.github.io/rust-clippy/rust-1.68.0/index.html#missing_safety_doc

Reply:
The `unsafe` functions here aren't public, so shouldn't be showing up in any docs - they're just internal helpers used to implement the public `append` method.